File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter14/PetzoldWindows8VectorDrawing/PetzoldWindows8VectorDrawing/RoundCappedLine.cpp
#include "pch.h"
#include "RoundCappedLine.h"

using namespace PetzoldWindows8VectorDrawing;

using namespace DirectX;
using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;

RoundCappedLine::RoundCappedLine()
{
    lineSegment1 = ref new LineSegment();
    arcSegment1 = ref new ArcSegment();
    lineSegment2 = ref new LineSegment();
    arcSegment2 = ref new ArcSegment();
}

void RoundCappedLine::Set(Point point1, Point point2, float radius)
{
    XMVECTOR pt1 = XMVectorSet(point1.X, point1.Y, 0, 0);
    XMVECTOR pt2 = XMVectorSet(point2.X, point2.Y, 0, 0);
    XMVECTOR vector = pt2 - pt1;
    XMVECTOR orthogonal = radius * XMVector2Orthogonal(
                                   XMVector3Normalize(vector));

    XMVECTOR pt1a = pt1 - orthogonal;
    XMVECTOR pt2a = pt1a + vector;
    XMVECTOR pt1b = pt1 + orthogonal;
    XMVECTOR pt2b = pt1b + vector;

    Point point1a(XMVectorGetX(pt1a), XMVectorGetY(pt1a));
    Point point2a(XMVectorGetX(pt2a), XMVectorGetY(pt2a));
    Point point1b(XMVectorGetX(pt1b), XMVectorGetY(pt1b));
    Point point2b(XMVectorGetX(pt2b), XMVectorGetY(pt2b));

    lineSegment1->Set(point1a, point2a);
    arcSegment1->Set(point2, radius, point2a, point2b);
    lineSegment2->Set(point2b, point1b);
    arcSegment2->Set(point1, radius, point1b, point1a);
}

void RoundCappedLine::GetAllX(float y, IVector<float>^ xCollection)
{
    arcSegment1->GetAllX(y, xCollection);
    lineSegment1->GetAllX(y, xCollection);
    arcSegment2->GetAllX(y, xCollection);
    lineSegment2->GetAllX(y, xCollection);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter14/Petzold.Windows8.VectorDrawing/Petzold.Windows8.VectorDrawing/RoundCappedLine.cs
using System.Collections.Generic;
using Windows.Foundation;

namespace Petzold.Windows8.VectorDrawing
{
    public struct RoundCappedLine : IGeometrySegment
    {
        LineSegment lineSegment1;
        ArcSegment arcSegment1;
        LineSegment lineSegment2;
        ArcSegment arcSegment2;

        public RoundCappedLine(Point point1, Point point2, double radius) : this()
        {
            Vector2 vector = new Vector2(point2 - new Vector2(point1));
            Vector2 normVect = vector;
            normVect = normVect.Normalized;

            Point pt1a = point1 + radius * new Vector2(normVect.Y, -normVect.X);
            Point pt2a = pt1a + vector;
            Point pt1b = point1 + radius * new Vector2(-normVect.Y, normVect.X);
            Point pt2b = pt1b + vector;

            lineSegment1 = new LineSegment(pt1a, pt2a);
            arcSegment1 = new ArcSegment(point2, radius, pt2a, pt2b);
            lineSegment2 = new LineSegment(pt2b, pt1b);
            arcSegment2 = new ArcSegment(point1, radius, pt1b, pt1a);
        }

        public void GetAllX(double y, IList<double> xCollection)
        {
            arcSegment1.GetAllX(y, xCollection);
            lineSegment1.GetAllX(y, xCollection);
            arcSegment2.GetAllX(y, xCollection);
            lineSegment2.GetAllX(y, xCollection);
        }
    }
}
